blocks/model.py

```diff
@@ -0,0 +1,181 @@
+        ----------
+        param_values : OrderedDict
+            Dictionary of (parameter name, :class:`~numpy.ndarray`) pairs.
+
+        """
+        params = self.get_params()
+
+        unknown = set(param_values) - set(params)
+        missing = set(params) - set(param_values)
+        if len(unknown):
+            logger.error("unknown parameter names:\n".format(unknown))
+        if len(missing):
+            logger.error("missing values for parameters:\n".format(missing))
+
+        for name, value in param_values.items():
+            if name in params:
+                params[name].set_value(value)
+
+    @abstractmethod
+    def get_criterion(self):
+        """Return the optimization criterion."""
+        pass
+
+    @abstractmethod
+    def get_top_bricks(self):
+        """Return the top-level bricks that are used in the model.
+
+        Returns
+        -------
+        bricks : list
+            List of bricks.
+
+        """
+        pass
+
+
+class Model(AbstractModel, ComputationGraph):
+    """Wraps a computation graph to support model interface.
+
+    This model covers the most common case when all information
+    about the model is contained in an annotated computation graph:
+    parameters are identified by the roles, bricks found by annotations.
+    Due to frequency of this case this class is called simply 'Model'
+    and not 'ComputationGraphModel'.
+
+    Parameters
+    ----------
+    outputs : (list of) :class:`~theano.Variable`
+        The output variables of the computation graph.
+
+    .. todo::
+
+        Overriding the automatically found parameters and bricks might
+        be needed.
+
+        If there top bricks in scan inner graphs, those will not be found.
+
+    """
+    def __init__(self, outputs):
+        super(Model, self).__init__(outputs)
+
+        bricks = [get_brick(var) for var in self if get_brick(var)]
+        children = set(chain(*(brick.children for brick in bricks)))
+        # Quadratic complexity: we should not have thousands of
+        # top-level bricks.
+        self.top_bricks = []
+        for brick in bricks:
+            if brick not in children and brick not in self.top_bricks:
+                self.top_bricks.append(brick)
+
+        brick_param_names = {
+            v: k for k, v in Selector(self.top_bricks).get_params().items()}
+        self.params = []
+        for param in VariableFilter(roles=[PARAMETER])(self.shared_variables):
+            if param in brick_param_names:
+                self.params.append((brick_param_names[param], param))
+            else:
+                self.params.append((param.name, param))
+        self.params = OrderedDict(self.params)
+
+    def get_criterion(self):
+        """Return the output variable, if there is a single one.
+
+        If there is only one output variable, it is a reasonable default
+        setting to assume that it is the optimization objective.
+
+        """
+        if len(self.outputs) == 1:
```

Comment:
Mm, I see, but maybe we should let the user set the cost as an optional keyword argument instead? It's just kind of strange to support multiple outputs, but not support retrieving the cost in that case. Right now the user would be forced to implement an entirely new `Model` class just if he wants to have two outputs of which the first one is the cost. It's also awkward that you can still instantiate the model, but that it will crash only when you do `get_criterion`.

You could do something like:

```python
"""
Parameters
------------
outputs : list
    A list of :class:`~tensor.TensorVariable` representing the outputs of your model.
cost : :class:`~tensor.TensorVariable`, optional
    If not given, `outputs` is expected to have only a single element (the cost). If
    outputs has multiple elements, you must pass the cost separately. If the cost
    is not passed as part of outputs, it will be added to it.
"""
def __init__(self, outputs, cost=None):
    if len(pack(outputs)) > 1 and cost is None:
        raise ValueError
    if cost is not None:
        if cost not in outputs:
            outputs.append(cost)
        self.cost = cost
    else:
        self.cost = unpack(outputs)
    self.outputs = outputs
```